---
title: "Pull Macosko et al. single cell mouse retina data"
output:
  html_notebook:
author: David McGaughey
date: 2017-09-05
---

David,
Please see the link below for the paper on Single-Cell retinal sequencing paper.  We can ask the following questions.
Whether Nlz2 present one or more clusters?
Whether Nlz2 present in Amacrine or ganglion cell clusters?
If present, what other transcriptional factors present in those clusters?

http://www.sciencedirect.com/science/article/pii/S0092867415005498

There is another paper on Single-cell sequencing done on Bipolar cell.  See the link below.
http://www.sciencedirect.com/science/article/pii/S0092867416310078

Thanks for your help.

Elan


https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE63472
https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE63472&format=file&file=GSE63472%5FP14Retina%5FlogDGE%2Etxt%2Egz
https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE63472&format=file&file=GSE63472%5FP14Retina%5Fmerged%5Fdigital%5Fexpression%2Etxt%2Egz
http://mccarrolllab.com/wp-content/uploads/2015/05/retina_clusteridentities.txt

Three approaches tried:

1. Seurat
2. Monocle
3. Aaron Lun (https://bioconductor.org/help/workflows/simpleSingleCell/)

Seurat was the more performant option, compared to Monocle. I also had better success with clustering assignments and t-SNE. 

Seurat-based approach
```{r}
library(cowplot)
library(Seurat)
library(tidyverse)
library(here)
# load processed data from scripts/process_macosko.R
load(here('data/retina_seurat_subSet.Rdata'))
# Choose gene outliers on mean-variability plot
retina <- FindVariableGenes(object = retina, x.low.cutoff = 0, y.cutoff = 2)
length(x = retina@var.genes)

# elbow at 30
PCElbowPlot(object = retina, num.pc = 40)
PrintPCA(object = retina, pcs.print = 1:36)
PCHeatmap(object = retina, pc.use = 1:12,100)
#PCHeatmap(object = retina, pc.use = 13:24,100)
#PCHeatmap(object = retina, pc.use = 25:36,100)

VizPCA(object = retina, pcs.use = 1:2)

TSNEPlot(object = retina, do.label = TRUE)
```

Print clusters with  Macosko et al. assignments
```{r}


#print with macosko clustering
TSNEPlot(object = retina, do.label = TRUE, group.by = 'Macosko_Clusters')

```

Plot marker genes and ZFP503.
```{r, fig.width=6, fig.height=4}
# plot individual genes
FeaturePlot(object = retina, features.plot = c("PAX6", "VSX2", "OPN1MW", "RLBP1","NRL","SLC17A6","CRX","ZFP503"), 
            cols.use = c("gray", 
                         "blue"))
```



Cells with ZFP503 Expression. 

`r table(retina@data['ZFP503',] > 0.1)['TRUE']` cells out of `r ncol(retina@data)` have detectable ZFP503 expression. 
```{r}
table(retina@data['ZFP503',] > 0.1)
```

Expression of key retinal markers in ZFP503 positive cells
```{r}
zfp503_pos_cells <- colnames(retina@data)[retina@data['ZFP503',] > 0.1]

retina@data[c("POU4F1", "CALB2", "ISL1", "LHX2","PAX6", "VSX2", "OPN1MW", "RLBP1","NRL","SLC17A6","CRX","ZFP503"), zfp503_pos_cells] %>% as.matrix() %>% data.frame() %>% rownames_to_column('Gene') %>% gather('cell_id', 'Expression', -Gene) %>% ggplot(aes(x=Gene, y=Expression)) + geom_violin() + geom_jitter(alpha=0.2) + theme(axis.text.x = element_text(angle = 90, hjust = 1))

retina@data[c("POU4F1", "CALB2", "ISL1", "LHX2","PAX6", "VSX2", "OPN1MW", "RLBP1","NRL","SLC17A6","CRX","ZFP503"), zfp503_pos_cells] %>% as.matrix() %>% data.frame() %>% rownames_to_column('Gene') %>% gather('cell_id', 'Expression', -Gene) %>% ggplot(aes(x=Gene, y=Expression)) + geom_boxplot() + geom_jitter(alpha=0.2) + theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust=0.5))
```

Macosko Clustering cell type assignments for ZFP503

Over half are assigned as amacrine
```{r}
retina@meta.data[zfp503_pos_cells,'Cell Type'] %>% table()
```

Most expressed genes across the `r table(retina@data['ZFP503',] > 0.1)['TRUE']` cells
```{r}
retina@data[,zfp503_pos_cells] %>% 
  as.matrix() %>% 
  data.frame() %>% 
  rownames_to_column('Gene') %>% 
  gather('cell_id', 'Expression', -Gene) %>% 
  group_by(cell_id) %>% 
  top_n(10) %>% 
  pull(Gene) %>% 
  table() %>% 
  sort(decreasing = T)
```

Find differentially expressed genes in ZFP503 positive cells relative to non-ZFP503 expressing cells
```{r}
retina_zfp503 <- retina
zfp503_tf <- retina@data['ZFP503',] > 0.1
zfp503_tf <- gsub(TRUE, 1, zfp503_tf)
zfp503_tf <- gsub(FALSE, 0, zfp503_tf)
names(zfp503_tf) <- names(retina@ident)
zfp503_tf <- as.factor(zfp503_tf)
retina_zfp503@ident <- zfp503_tf

zfp503_markers <- FindMarkers(object = retina_zfp503, ident.1 = 1)

zfp503_markers %>% DT::datatable()

```
